dashboard/modules/log/log_agent.py

```diff
@@ -252,6 +289,79 @@ async def ListLogs(self, request, context):
             log_files.append(p.name)
         return reporter_pb2.ListLogsReply(log_files=log_files)
 
+    @classmethod
+    async def _find_task_log_offsets(
+        cls, task_id: str, attempt_number: int, lines: int, f: io.BufferedIOBase
+    ) -> Tuple[int, int]:
+        """Find the start and end offsets in the log file for a task attempt
+        Current task log is in the format of below:
+
+            :job_id:xxx
+            :task_name:xxx
+            :task_attempt_start:<task_id>-<attempt_number>
+            ...
+            actual user logs
+            ...
+            :task_attempt_end:<task_id>-<attempt_number>
+            ... (other tasks)
+
+
+        For async actor tasks, task logs from multiple tasks might however
+        be interleaved.
+        """
+
+        # Find start
+        task_attempt_start_magic_line = (
+            f"{LOG_PREFIX_TASK_ATTEMPT_START}{task_id}-{attempt_number}\n"
+        )
+
+        # Offload the heavy IO CPU work to a thread pool to avoid blocking the
+        # event loop for concurrent requests.
+        with concurrent.futures.ThreadPoolExecutor() as pool:
```

Comment:
You want to specify num_workers here. The default is num_cpus/2 or sth like that which is pretty large IIRC

Reply:
Also consider maintaing a single pool instead of keep creating a new one?